File: GenomicRangeQuery/C/GenomicRangeQuery.c
// General understanding: We compare S from P[x] to Q[x]; x++, and find smallest value "genome"
// A = 1
// C = 2
// G = 3
// T = 4

// you can write to stdout for debugging purposes, e.g.
// printf("this is a debug message\n");
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct Results solution(char *S, int P[], int Q[], int M)
{
    unsigned length;
    length = strlen(S);

    int A[length];
    int C[length];
    int G[length];
    int T[length];

    struct Results result = {calloc(M, sizeof(int)), M};

    int i = 0;

    while (i < length) 
    {
        if (i == 0)
            A[i] = C[i] = G[i] = T[i] = 0;
        else 
        {
            A[i] = A[i - 1];
            C[i] = C[i - 1];
            G[i] = G[i - 1];
            T[i] = T[i - 1];
        }
        switch (S[i]) 
        {
            case 'A':
                ++A[i];
                break;
            case 'C':
                ++C[i];
                break;
            case 'G':
                ++G[i];
                break;
            case 'T':
                ++T[i];
                break;
        }
        i++;
    }

    i = 0;

    // this is so not my doings, but ill try to explain what i got from this:
    // we compare which of the results changes in give time period from our array
    while (i < M) 
    {
        if (A[Q[i]] - (P[i] > 0 ? A[P[i] - 1] : 0) > 0)
            result.A[i] = 1;
        else if (C[Q[i]] - (P[i] > 0 ? C[P[i] - 1] : 0) > 0)
            result.A[i] = 2;
        else if (G[Q[i]] - (P[i] > 0 ? G[P[i] - 1] : 0) > 0)
            result.A[i] = 3;
        else
            result.A[i] = 4;
        i++;
    }
    return result;
}

File: GenomicRangeQuery/Python/GenomicRangeQuery.py
# https://codesays.com/2014/solution-to-genomic-range-query-by-codility/

# there was so much to learn about effiency. i hope everything is now imprinted on my brain :D

# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

# okay so this approach divided:
# we have Cost func to return numeric value for genome in hand
# we have arra[] that stores when the nuc was last found
# and we have lastFound[] that has values for found nuc's (if its over -1, there is nuc of that value)
# arra[i][j] = k
# -> i: type of nucleoid, [0] = A, [1] = C, [2] = G and [3] = T
# -> j: this nucleoid
# -> K: appears in this position (including j)
# so if k == -1, it means there are no this type of nucs


# small helper to return cost of the current "genome"
def cost(S):
    if S == 'A':
        return 1
    elif S == 'C':
        return 2
    elif S == 'G':
        return 3
    else:
        return 4

def solution(S, P, Q):
    # init empty array, and last found array for A, C, G and T, defaulted to -1
    # also make empty array for upcoming result
    arra = []
    result = []
    lastFound = [-1, -1, -1, -1]
    for i in range(0, len(S)) :
        lastFound[cost(S[i]) - 1] = i
        arra.append(list(lastFound))
    # loop arrays from 0 -> len(P)
    # then check by slices of 4 (since there is 4 different genomes in array), if there is bigger genomes on our made array and we add them on our result
    for i in range(0, len(P)) :
        for j in range(0, 4) :
            if arra[Q[i]][j] >= P[i] :
                result.append(j + 1)
                break
    return result